### PR TITLE
(det)shuffle shardlist before (node/worker) splitting in WebDataset

### PR DESCRIPTION
**Pull request by jrcavani**
In distributed training with GPU processes and data loader workers, one common pattern is shuffling the webdataset shards in an identical way across all Python processes, and split based on some consistent rule to have non-overlapping data. This is the strategy [`DistributedSampler`](https://pytorch.org/docs/stable/_modules/torch/utils/data/distributed.html#DistributedSampler) adopts. We can achieve this with the v2 DataPipeline API:

```python
    dataset = wds.DataPipeline(
        wds.SimpleShardList("source-{000000..00007}.tar"),
        wds.detshuffle(),
        wds.split_by_node,
        wds.split_by_worker,
    )
```

```
rank: 0, world_size: 8, {'url': 'source-000006.tar'}
rank: 1, world_size: 8, {'url': 'source-000007.tar'}
rank: 5, world_size: 8, {'url': 'source-000005.tar'}
rank: 2, world_size: 8, {'url': 'source-000003.tar'}
rank: 4, world_size: 8, {'url': 'source-000002.tar'}
rank: 3, world_size: 8, {'url': 'source-000000.tar'}
rank: 6, world_size: 8, {'url': 'source-000001.tar'}
rank: 7, world_size: 8, {'url': 'source-000004.tar'}
```

However, the current `WebDataset` API puts shuffling logic **after** the splitting logic, so shuffling only takes place within each data loader worker on each rank.

```python
    dataset = (
        wds.WebDataset("source-{000000..00007}.tar", shardshuffle=True, detshuffle=True, nodesplitter=wds.split_by_node)
    )
    dataset.pipeline.pop()
```

```
rank: 1, world_size: 8, {'url': 'source-000001.tar'}
rank: 5, world_size: 8, {'url': 'source-000005.tar'}
rank: 2, world_size: 8, {'url': 'source-000002.tar'}
rank: 4, world_size: 8, {'url': 'source-000004.tar'}
rank: 0, world_size: 8, {'url': 'source-000000.tar'}
rank: 3, world_size: 8, {'url': 'source-000003.tar'}
rank: 6, world_size: 8, {'url': 'source-000006.tar'}
rank: 7, world_size: 8, {'url': 'source-000007.tar'}
```

This is not necessarily bad. Thinking in favor of the existing approach, the GPU processes are going through samples in parallel, and it shouldn't really matter each specific GPU process gets a different shard across the same group of shards. In addition, having fixed set of shards for each rank encourages data caching during multi-node training, where each rank always gets the same set of shards, and perform shard shuffling if there are more than 1 item in the set of shards.

It's just a little unexpected during testing, because PyTorch has long been using this format, and the [ordering of the named params](https://github.com/webdataset/webdataset/blob/main/webdataset/compat.py#L95) - `shardshuffle` and `detshuffle` preceed `nodesplitter`, suggesting the same PyTorch-like behavior. So if it is desired to keep the existing behavior, it might help change the param order, and update the docs.

### CODE REVIEW

**Comment by jrcavani** [comment]
Just to note, trying to emulate PyTorch's `DistributedSampler` with `detshuffle` also won't actually shuffle between epochs, within the same rank, when using `num_workers > 1`. It works if one uses `persistent_worker=True`, because of how `detshuffle()` internally keeps tracking of the epoch increment, and the increment within the forked data loader workers won't persist in between epochs.

```python
    dataset = wds.DataPipeline(
        wds.SimpleShardList("source-{000000..00015}.tar"),
        wds.detshuffle(),
        wds.split_by_node,
        wds.split_by_worker,
    )
```

```
rank: 7, {'url': 'source-000013.tar'}
rank: 5, {'url': 'source-000008.tar'}
rank: 5, {'url': 'source-000005.tar'}
rank: 2, {'url': 'source-000006.tar'}
rank: 3, {'url': 'source-000000.tar'}
rank: 0, {'url': 'source-000012.tar'}
rank: 0, {'url': 'source-000011.tar'}
rank: 6, {'url': 'source-000007.tar'}
rank: 4, {'url': 'source-000004.tar'}
rank: 4, {'url': 'source-000001.tar'}
rank: 6, {'url': 'source-000014.tar'}
rank: 1, {'url': 'source-000015.tar'}
rank: 1, {'url': 'source-000003.tar'}
rank: 2, {'url': 'source-000002.tar'}
rank: 7, {'url': 'source-000010.tar'}
rank: 3, {'url': 'source-000009.tar'}
--------------------------------------
(identical as above)
rank: 5, {'url': 'source-000008.tar'}
rank: 5, {'url': 'source-000005.tar'}
rank: 1, {'url': 'source-000015.tar'}
rank: 7, {'url': 'source-000013.tar'}
rank: 7, {'url': 'source-000010.tar'}
rank: 1, {'url': 'source-000003.tar'}
rank: 6, {'url': 'source-000007.tar'}
rank: 6, {'url': 'source-000014.tar'}
rank: 4, {'url': 'source-000004.tar'}
rank: 3, {'url': 'source-000000.tar'}
rank: 3, {'url': 'source-000009.tar'}
rank: 2, {'url': 'source-000006.tar'}
rank: 2, {'url': 'source-000002.tar'}
rank: 4, {'url': 'source-000001.tar'}
rank: 0, {'url': 'source-000012.tar'}
rank: 0, {'url': 'source-000011.tar'}

```

A workaround without `persistent_worker=True` is to manually set the epoch for 
```
    for epoch in range(2):
        for idx, i in enumerate(dataloader):
            print(f"rank: {rank}, {i}")

        if rank == 0:
            print("--------------------------------------")

        dataset.stage(1).epoch = epoch
```

```
rank: 7, {'url': 'source-000013.tar'}
rank: 1, {'url': 'source-000015.tar'}
rank: 1, {'url': 'source-000003.tar'}
rank: 2, {'url': 'source-000006.tar'}
rank: 2, {'url': 'source-000002.tar'}
rank: 7, {'url': 'source-000010.tar'}
rank: 3, {'url': 'source-000000.tar'}
rank: 5, {'url': 'source-000008.tar'}
rank: 5, {'url': 'source-000005.tar'}
rank: 0, {'url': 'source-000012.tar'}
rank: 0, {'url': 'source-000011.tar'}
rank: 4, {'url': 'source-000004.tar'}
rank: 6, {'url': 'source-000007.tar'}
rank: 4, {'url': 'source-000001.tar'}
rank: 3, {'url': 'source-000009.tar'}
rank: 6, {'url': 'source-000014.tar'}
--------------------------------------
(all different but still unique shards for each rank)
rank: 1, {'url': 'source-000009.tar'}
rank: 1, {'url': 'source-000003.tar'}
rank: 2, {'url': 'source-000013.tar'}
rank: 2, {'url': 'source-000005.tar'}
rank: 7, {'url': 'source-000007.tar'}
rank: 7, {'url': 'source-000002.tar'}
rank: 5, {'url': 'source-000015.tar'}
rank: 5, {'url': 'source-000000.tar'}
rank: 6, {'url': 'source-000011.tar'}
rank: 6, {'url': 'source-000010.tar'}
rank: 0, {'url': 'source-000004.tar'}
rank: 0, {'url': 'source-000008.tar'}
rank: 4, {'url': 'source-000001.tar'}
rank: 4, {'url': 'source-000014.tar'}
rank: 3, {'url': 'source-000012.tar'}
rank: 3, {'url': 'source-000006.tar'}
```

**Comment by tmbdev** [comment]
Thank you. Yes, the idea was to facilitate caching on different nodes by assigning the same shards to each during each epoch. I will look into this.

I should add that going forward, the new "wids" subpackage provides an alternative for distributed training and uses distributed samplers for this purpose.

**Comment by rwightman** [comment]
@jrcavani FWIW, this a that I hacked together sometime ago for OpenCLIP addresses some of the concerns here, places splits after detshuffle, but also customizes detshuffle and ResampledShards behaviour to work around some seeding / epoch gotchas for distributed training, persistent workers. It's been used quite a bit and forked to a number of other projects at this point...

https://github.com/mlfoundations/open_clip/blob/main/src/training/data.py#L328-L443

**Comment by jrcavani** [comment]
Thank you for the link @rwightman . @tmbdev I will close this PR then.
